core/src/main/java/org/springframework/security/authorization/AuthorizationManager.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.authorization;
+
+import java.util.function.Supplier;
+
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.core.Authentication;
+
+/**
+ * An Authorization manager which can determine if an {@link Authentication} has access to
+ * a specific object.
+ *
+ * @param <T> the type of object that the authorization check is being done one.
+ * @author Evgeniy Cheban
+ */
+@FunctionalInterface
+public interface AuthorizationManager<T> {
+
+	/**
+	 * Determines if access is granted for a specific authentication and object.
+	 * @param authentication the Authentication to check
+	 * @param object the object to check
+	 * @return a decision
+	 */
+	AuthorizationDecision check(Supplier<Authentication> authentication, T object);
```

Comment:
Maybe it would be better to return an empty `Optional` instead of `null`?

Reply:
`Optional` adds some extra overhead; it's been observed to cause excess GC at the framework level. While `Optional` is nice, since this could potentially be called on every request, I'd prefer `null`.